[Fast text codepath] CharactersTreatedAsSpace is overly prescriptive

litherum opened the pull request:
#### 6566fb8f9216aae136892c5f32e6e50367de2969
<pre>
[Fast text codepath] CharactersTreatedAsSpace is overly prescriptive
<a href="https://bugs.webkit.org/show_bug.cgi?id=260843">https://bugs.webkit.org/show_bug.cgi?id=260843</a>
rdar://114607926

Reviewed by Cameron McCormack.

When we lay out web content, we make multiple passes over the text,
chopping it up multiple different ways. Because of this, we have an
invariant that the width of a space can&apos;t change depending on the
bounds of how we&apos;re chopping up the text. Kerning can affect the
width of a space, so we enforce the invariant by having a postprocess
that we perform after shaping to retroactively go back and adjust
the width of any glyphs associated with space characters, to reset
their widths. This is called &quot;CharactersTreatedAsSpace.&quot;

However, CharactersTreatedAsSpace actually does something else, too:
it also adjusts the width of the character directly before the space
to be whatever it was before shaping. However, this is kind of
bogus - it&apos;s totally valid for shaping to adjust the width of
whichever character happens to be before the space.

Now that we&apos;re trying to make the fast text codepath handle complex
character clusters, we&apos;re hitting the case more often where shaping
is adjusting the non-space glyphs to be correct, but then our
postprocess is coming along and messing them up. This patch removes
that logic, just for the glyphs before the space characters. The
logic about trying to reset the space characters&apos; width is still
there.

* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::OriginalAdvancesForCharacterTreatedAsSpace::OriginalAdvancesForCharacterTreatedAsSpace):
(WebCore::WidthIterator::applyFontTransforms):
(WebCore::WidthIterator::advanceInternal):

Canonical link: <a href="https://commits.webkit.org/267541@main">https://commits.webkit.org/267541@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/871afdb29f6c7f33cd2bccc6b7c555d9a9c481ce

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16885 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17208 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17665 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18668 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15808 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17075 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20440 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17346 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18067 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17084 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17442 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/14615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19475 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14685 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15307 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/15673 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15474 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19790 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16084 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15251 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4049 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19613 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15914 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/platform/graphics/WidthIterator.cpp:
```diff
@@ -62,18 +62,16 @@ WidthIterator::WidthIterator(const FontCascade& font, const TextRun& run, HashSe
 
 struct OriginalAdvancesForCharacterTreatedAsSpace {
-    explicit OriginalAdvancesForCharacterTreatedAsSpace(GlyphBufferStringOffset stringOffset, bool isSpace, float advanceBefore, float advanceAt)
+    explicit OriginalAdvancesForCharacterTreatedAsSpace(GlyphBufferStringOffset stringOffset, bool isSpace, float advanceAt)
```
Maybe rename "advanceAt" (and "advanceAtCharacter") to just "advance"? Since we no longer have two different advances.

Also the class name could drop the "s" in "OriginalAdvances".

heycam reviewed (approved):
s/with of a space/width of a space/ in the commit message.

litherum committed:
[Fast text codepath] CharactersTreatedAsSpace is overly prescriptive
https://bugs.webkit.org/show_bug.cgi?id=260843
rdar://114607926

Reviewed by Cameron McCormack.

When we lay out web content, we make multiple passes over the text,
chopping it up multiple different ways. Because of this, we have an
invariant that the width of a space can't change depending on the
bounds of how we're chopping up the text. Kerning can affect the
width of a space, so we enforce the invariant by having a postprocess
that we perform after shaping to retroactively go back and adjust
the width of any glyphs associated with space characters, to reset
their widths. This is called "CharactersTreatedAsSpace."

However, CharactersTreatedAsSpace actually does something else, too:
it also adjusts the width of the character directly before the space
to be whatever it was before shaping. However, this is kind of
bogus - it's totally valid for shaping to adjust the width of
whichever character happens to be before the space.

Now that we're trying to make the fast text codepath handle complex
character clusters, we're hitting the case more often where shaping
is adjusting the non-space glyphs to be correct, but then our
postprocess is coming along and messing them up. This patch removes
that logic, just for the glyphs before the space characters. The
logic about trying to reset the space characters' width is still
there.

* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::OriginalAdvancesForCharacterTreatedAsSpace::OriginalAdvancesForCharacterTreatedAsSpace):
(WebCore::WidthIterator::applyFontTransforms):
(WebCore::WidthIterator::advanceInternal):

Canonical link: https://commits.webkit.org/267541@main
